autoextract_poet/items.py

```diff
@@ -7,12 +7,45 @@
 class Item:
 
     @classmethod
-    def from_dict(cls, item: Optional[Dict]):
-        return cls(**item) if item else None  # type: ignore
+    def from_dict(cls, item: dict):
+        return cls(**item)  # type: ignore
 
     @classmethod
-    def from_list(cls, items: List[Dict]):
-        return [cls.from_dict(item) for item in items if item]
+    def _get_attr_cls(cls, annotation):
+        """Inspect type annotation searching for a subclass of Item"""
+        if type(annotation) is type:
+            # Annotation is a class/type
+            return annotation if issubclass(annotation, Item) else None
+
+        # Annotation is probably a typing instance
+        for arg in annotation.__args__:
+            attr_cls = cls._get_attr_cls(arg)
+            if attr_cls:
+                return attr_cls
+
+    def __attrs_post_init__(self):
+        for attribute in self.__attrs_attrs__:
+            annotation = self.__annotations__.get(attribute.name)
+            attr_cls = self._get_attr_cls(annotation)
+            if not attr_cls:
+                # Annotation does not mention an Item subclass
+                continue
+
+            def get_new_value(value):
+                if isinstance(value, dict):
+                    # We can build an instance from a dict
+                    return attr_cls.from_dict(value)
+                else:
+                    # Nothing to do here
+                    return value
+
+            value = getattr(self, attribute.name, None)
+            if isinstance(value, (list, tuple, )):
```

Comment:
Here if the attribute expects a list or a tuple is deducted from the value, but I think it should be deducted from the annotation. This way it can be detected if there is an unexpected value and raise a meaningful error.

Reply:
Sure, I went on this path because it was easy to cover our current schemas. Maybe we should be using something else other than `attrs` like for example `Pydantic` in order to have proper schema validation and to avoid having to write our own conversion logic.